### PR TITLE
TICK-358 longitudinal features lagger

### PR DESCRIPTION
**Pull request by MaryanMorel**
Add a preprocessor to lag exposure features matrices

### CODE REVIEW

File: tick/preprocessing/longitudinal_features_lagger.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        """Fit the feature lagger using the features matrices list.
+
+        Parameters
+        ----------
+        X : list of np.ndarray or list of scipy.sparse.csr_matrix,
+            list of length n_samples, each element of the list of 
+            shape=(n_intervals, n_features)
+            The list of features matrices.
+
+
```
Comment (stephanegaiffas) [inline tick/preprocessing/longitudinal_features_lagger.py]: Single line between arguments in docstrings

File: tick/preprocessing/longitudinal_features_lagger.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+            The censoring data. This array should contain integers in 
+            [1, n_intervals]. If the value i is equal to n_intervals, then there
+            is no censoring for sample i. If censoring = c < n_intervals, then 
+            the observation of sample i is stopped at interval c, that is, the 
+            row c - 1 of the correponding matrix. The last n_intervals - c rows
+            are then set to 0.
+
+        Returns
+        -------
+        output : `LongitudinalFeaturesProduct`
```
Comment (stephanegaiffas) [inline tick/preprocessing/longitudinal_features_lagger.py]: LongitudinalFeaturesLagger

File: tick/preprocessing/longitudinal_features_lagger.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+            The fitted current instance.
+        """
+        self._reset()
+        base_shape = X[0].shape
+        X = LFP._check_X_consistency(X, base_shape, "float64")
```
Comment (stephanegaiffas) [inline tick/preprocessing/longitudinal_features_lagger.py]: check_X_consistency should be in a tick.preprocessing.utils script.
But we can leave this like that for now

File: tick/preprocessing/longitudinal_features_lagger.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+        "_n_init_features": {"writable": False},
+        "_n_output_features": {"writable": False},
+        "_n_intervals": {"writable": False},
+        "_cpp_preprocessor": {"writable": False},
+        "_fitted": {"writable": False}
+    }
+
+    def __init__(self, n_lags=0, n_jobs=-1):
+        Base.__init__(self)
+        self.n_lags = n_lags
```
Comment (stephanegaiffas) [inline tick/preprocessing/longitudinal_features_lagger.py]: Don't we need some tests about n_lags ? n_lags >= 0 and one checking that n_lags is not too large ?

File: tick/preprocessing/longitudinal_features_lagger.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+            The censoring data. This array should contain integers in 
+            [1, n_intervals]. If the value i is equal to n_intervals, then there
+            is no censoring for sample i. If censoring = c < n_intervals, then 
+            the observation of sample i is stopped at interval c, that is, the 
+            row c - 1 of the correponding matrix. The last n_intervals - c rows
+            are then set to 0.
+
+        Returns
+        -------
+        output : `[np.ndarrays]`  or `[csr_matrices]`, shape=(n_intervals, n_features)
```
Comment (stephanegaiffas) [inline tick/preprocessing/longitudinal_features_lagger.py]: I prefer `list` of `numpy.ndarray` or `csr_matrix`, but that's a detail

File: tick/preprocessing/longitudinal_features_lagger.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+            The list of features matrices with added lagged features.
+        """
+
+        n_samples = len(X)
+        censoring = self._check_censoring_consistency(censoring, n_samples)
+        base_shape = (self._n_intervals, self._n_init_features)
+        X = LFP._check_X_consistency(X, base_shape, "float64")
+        if sps.issparse(X[0]):
+            X_with_lags = [self._sparse_lagger(x, int(censoring[i]))
+                           for i, x in enumerate(X)]  # Don't get why int() is required here as censoring_i is uint64
```
Comment (stephanegaiffas) [inline tick/preprocessing/longitudinal_features_lagger.py]: That is weird indeed... but if it's working, let's do it like that :)

File: tick/preprocessing/longitudinal_features_lagger.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+        else:
+            X_with_lags = [self._dense_lagger(x, int(censoring[i]))
+                           for i, x in enumerate(X)]
+
+        return X_with_lags
+
```
Comment (stephanegaiffas) [inline tick/preprocessing/longitudinal_features_lagger.py]: There's something I don't get. It is supposed to add the product features to the given list X, but I don't see where the passed list is extended with the products ? It seems that you construct the products, and return the list, without keeping the original ones ??

Comment (MaryanMorel) [inline tick/preprocessing/longitudinal_features_lagger.py]: It adds lags, i.e. add new columns to each element of the list. These elements are ndarrays or csr_matrices, so it seems as inefficient to rebuild the objects than to extend them: ndarrays will trigger copies at some point to be contiguous in memory, and it cannot be done with scipy.sparse.csr_matrices

Comment (stephanegaiffas) [inline tick/preprocessing/longitudinal_features_lagger.py]: Oh okay then... that's sure is the best strategy

**Comment by Mbompr** [comment]
@MaryanMorel The two commits should have been squashed
https://twitter.com/jamesfublo/status/402407321265274881?ref_src=twsrc%5Etfw&ref_url=http%3A%2F%2Fjamescooke.info%2Fgit-to-squash-or-not-to-squash.html
Also each commit should contain the task name it is related to :)

**Comment by MaryanMorel** [comment]
I know, my bad… I saw that when it was too late, and I didn’t want to mess with the master to make it right :(

Maryan MOREL
Centre de Mathématiques Appliquées (CMAP) <http://www.cmap.polytechnique.fr/> 
École Polytechnique 
91128 PALAISEAU CEDEX
+33 (0)1 69 33 45 80

**Pull request merged**